### ISSUE TITLE
Unclear definition of "weighting" in ramp_fitting step from docs

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#5300]
_Issue [JP-1667](https://jira.stsci.edu/browse/JP-1667) was created on JIRA by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza):_

It is currently unclear from the documentation what the "weighting" parameter in the ramp_fitting step is doing. Naively, "weighting" might be associated with performing an estimate of the slope weighting for the errors, but that wouldn't formally be Ordinary Least Squares which is the default method when running the ramp_fitting step (ols).

This is important to understand because, in general, for exoplanet time-series observations it seems that setting "weighting = unweighted" gives much better results than the default "weighted = optimal" version. The attached slides show an example of this with a NIRCam simulation analyzed by [Nikolay Nikolov](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nnikolov).

Based on tests made in [this notebook](https://github.com/nespinoza/jwst_pipeline_testing/blob/master/Stage%201%20testing%20-%20NIRISS%20SOSS.ipynb) (which were made with NIRISS/SOSS simulations), my hunch is that the "weighting" parameter is just defining what to do in cases when there is a DQFLAG different than zero (with "unweighted" meaning nothing is done). Whatever the case, it might be perhaps optimal to have "weighted = unweighted" as the default for TSO observations, as with the typical small number of groups per integration (and hence, frame-times per group), it might be more likely that jumps are actually due to statistical fluctuations and not due to cosmic rays. We will discuss this anyways on our next TSO WG meeting before opening an "Improvement" ticket on this matter. We want this ticket to be solely about defining what the "weighting" parameter actually does.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=487236#comment-487236):_

I'm not clear on what the problem is here. Is this a documentation problem or are you requesting a change? Uniform weighting has a lower S/N noise than optimum when the observation is Poisson noise limited. When it is read noise limited, they give the same result.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=487399#comment-487399):_

Hi [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan); very sorry this is not clear from my original message/ticket. I initially strictly meant the documentation: the documentation should state, explicitly, what is done when "weighting=optimal" is set, and what is done when "weighting=unweighted" is set. This, in turn, might help us understand if this algorithm is optimal or not for Time Series Observations. This is explained (perhaps not in an "optimal" way!) in my first paragraph. The Read The Docs article (https://jwst-pipeline.readthedocs.io/en/latest/jwst/ramp_fitting/description.html) only talks about the variances of the slope, but it doesn't tell how the actual estimator of the slope looks like.

In my last paragraph I state that, if indeed the "weighting" parameter is performing using some kind of weighted least squares in some way, then the parameter names should be changed either for the "weighting" parameter, the "algorithm" parameter, of both. Weighted least squares is, by definition, not ordinary least squares. So, if indeed with "weighted = optimal" the algorithm is performing some form of weighted least squares, this is inconsistent with the "algorithm" variable being set to "ols" by default.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=487535#comment-487535):_

[Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) I have to say that I don't understand the attached plot. Was this run through the Caldetector1 pipeline? Something other than the weighting must have changed because the code only does optimal weighting. The input parameter is ignored (at least it is supposed to be, I'll check.). Even if the code is somehow being driven down the unweighted path, I can't see how the difference in the weighting could cause the bimodal distribution of the slopes.

As for the name of the parameters and routines, I've got a long list of things I'd like to rename too.

The primary distinction is between optimum weighting and generalized least squares(GLS). Karl Gordon has shown that GLS has several advantages but can be significantly slower and does not currently handle some of the edge cases correctly. Though for pixels with a large fraction of cosmic rays it can actually be faster than OLS.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=487537#comment-487537):_

Hi [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) --- we can wait for [Nikolay Nikolov](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nnikolov) to give an explaination next week (he is on vacation this week), but my understanding is that yes, the was ran through Caldetector1, an the only difference was the weighting between both runs. From my own check of what "weighting=optimal" and "algorithm=ols" was doing, I actually saw that the slopes **estimators** (i.e., the equation of the slope given the data) were not doing any kind of weighting, other than just paying attention to the DQ flags, and then using different methods depending if the first, middle or last group in my 3-group integration test was identified as an outlier by the jump step. It was doing simple OLS, i.e., not weighting the datapoints at all (see the Notebook I linked above).

**Comment by hbushouse** [comment]
The weighting, at least theoretically, refers to the weighting applied to line fits within the OLS method, using the scheme developed years ago by Fixsen et al. (2000). In this scenario the weighting is signal-dependent, where pixels with low signal (countrate) get nearly uniform weighting for each group in the segment being fitted, gradually varying to the case of pixels with very high signal (countrate), which essentially give all the weight to only the first and last groups within the segment. That's the weighting used when computing the slope for an individual ramp segment, i.e. the groups that occur in between CR hits. Yet another weighting scheme (just based on the computed noise model for the pixel) is used to then average the slopes for all segments to give a final slope for a given integration. The same type of weighting is again used when averaging together the slopes from all integrations (essentially averaging together - all at once - the slopes from every segment over all integrations).

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=487556#comment-487556):_

[Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) Looking at the notebook I think I see a problem with using this dataset to investigate the effects of weighting. The changes in the weighting are to lower the non-end point values compared to the end-points. Your data set has only three groups. So, the only group affected in the central group. Uniform and optimum will give the same slope for a 3-group integration. Actually, you can put any value or weight you want for the central value and the slope does not change. 

Below is a link to the soccer document I wrote a long time ago about optimum weighting. This lead to changes in the NICMOS pipeline.

<https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/legacy/nicmos/instrument-science-reports-isrs/_documents/isr_2008_002.pdf> 

The other problem is that the jump step is not designed to work with 3 group integrations. The minimum should be set to 5 (maybe 4 but I'd have to check). Three groups is two differences. There's no way for the current algorithm to tell which is a valid jump.  I thought I tested this case but I clearly didn't. I'll open an issue on Github.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=487966#comment-487966):_

[Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) Right; that is I think what is written in the Notebook as well. I believe that should be written explicitly in the ReadTheDocs. In general, *all* the decisions (e.g., dividing the slope in two, etc.) should be written over there. Otherwise it is just impossible to find out what is going on (in the notebook, I found out about all that via Slack, but this is not explicit in the docs).

As for the jump step not designed to work with 3 group ints, we briefly talked about this with [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on Slack but in regard to the slope fitting (my take was really that in 3 group integrations, it didn't make sense to divide the slopes because you are left with only one datapoint per side, which leads to bad results like the ones noted in the notebook as well --- see our discussion here: <https://stsci.slack.com/archives/CBQQMLA7Q/p1596060098059300).> 

And yes, the jumps step is very aggresive on small number of groups, we didn't know why and referred that to the reference files, but glad to hear it was not supposed to be ran for small number of groups!

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=489004#comment-489004):_

[Nikolay Nikolov](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nnikolov) [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) 

After the TSOWG decides how to proceed on this issue, it would probably still be good to file a bug ticket explaining the trouble with accessing the "weighting" parameter using the .call() method or the .run() method, since that's separate from the documentation issue above. 

Nikolay can correct me if this is not right, but I believe you can supply the "weighting = 'unweighted'" argument when running the ramp-fit step with the .run() method, but you are not able to supply the same argument using the .call() method (you get a "ValidationError: Extra value 'weighting' in root" message). 

I think this was probably just a bug from when the algorithm was updated a couple years ago. According to L29: <https://github.com/spacetelescope/jwst/blob/master/jwst/ramp_fitting/ramp_fit_step.py#L29>
```java
# As of 04/26/17, the only allowed algorithm is 'ols', and the 
# only allowed weighting is 'optimal'.
```
but "unweighted" is still allowed using the .run() method.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nikolay Nikolov](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nnikolov) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=489008#comment-489008):_

[Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) your comment above is correct.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=489039#comment-489039):_

I'm confused here but it's possible that I don't understand the way Python is setting class attributes. Weighting is a class attribute that is hard-coded to be 'optimal'. It could be possible that by setting it in the run method it overrides the hard-coded value (not really hard-coded, if you can change it.) 

Different steps handle the transfer of class attributes to local variables differently.  We should check that if this is true we don't allow it other places either.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nikolay Nikolov](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nnikolov) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=489054#comment-489054):_

To some extend it is good that the run() command still can do the 'unweighted' option. This is the only way to get the better result for transiting exoplanet TSOs.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=489070#comment-489070):_

Yeah, I don't understand how it can be affecting the slopes. Could you point me to the input uncal file? One thing you could try is to skip the jump step. It should not run with only three groups anyway.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=505329#comment-505329):_

[Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) [Nikolay Nikolov](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nnikolov) I see this ticket had some discussion in an October TSO WG meeting - is there any new information relevant to the issue? Is this ticket's issue strictly one of ramp_fit documentation for readthedocs?

**Comment by stscijgbot-jp** [comment]
_Comment by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=505351#comment-505351):_

Hi [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) --- no new information on this other than what was stressed in these comments. Indeed, this ticket is only about the readthedocs details. It is hard from the readthedocs to make our own slope fits match the ones from the pipeline with the different methods.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=505420#comment-505420):_

Again, the code is set to always use optimal weighting. It will always give a better estimate of the flux than unweighted. The math behind this is in Fixsen et al 2000 PASP 112, 1350 <https://arxiv.org/ct?url=https%3A%2F%2Fdx.doi.org%2F10.1086%2F316626&v=5dd17f42>

I simplified the weighting formula from Fixsen et al and we applied this to the NICMOS, WFC3, and the JW detectors. This is documented in Regan, JWST-STScI-001212 from 2007. <https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/legacy/nicmos/instrument-science-reports-isrs/_documents/isr_2008_002.pdf>

If you can show errors in Fixen's work, you're way ahead of me.

 
It seems like this can be closed by citing Fixsen et al and Regan 2007

**Comment by stscijgbot-jp** [comment]
_Comment by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=505432#comment-505432):_

Hi [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) --- I believe it all depends on how thorough we want the documentation to be. I think adding the weights that were used in the readthedocs *and* citing the papers might be the way to go.

N.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=506338#comment-506338):_

[Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan)
I have added the relevant equations, table and explanation from Fixsen/Regan to the docs, with the caveat that our code appears to interpolate P rather than bin to the six values provided by Fixsen. Please take a look at your convenience and let me know if this is:

 - helpful
 - clear
 - correct

Thanks for your time.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=506555#comment-506555):_

Fixed by #5682

**Comment by stscijgbot-jp** [comment]
_Comment by [Nestor Espinoza](https://jira.stsci.edu/secure/ViewProfile.jspa?name=nespinoza) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=514222#comment-514222):_

This is all great, thanks folks. One detail: is Reagan 2007 available for the community [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan)?

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-1667?focusedCommentId=514255#comment-514255):_

I added the interpolate to avoid the discontinuities at the different S/N boundaries. I don't remember taking it out but that doesn't mean I didn't do it. 

Well, JW is different than Hubble. For HST every technical report is public. For JW we put things in SOCCER which is not visible. I've complained about this for years but the CM folks don't care, the instrument teams have higher priorities, and the mission office doesn't want to confuse the public with obsolete reports.
